subsys/bluetooth/services/ots/ots_client.c

```diff
@@ -499,6 +499,7 @@ int bt_ots_client_unregister(uint8_t index)
 {
 	if (index < ARRAY_SIZE(otc_insts)) {
 		memset(&otc_insts[index], 0, sizeof(otc_insts[index]));
+		cur_inst = NULL;
```

Comment:
Shouldn't this only be cleared if `cur_inst == &otc_insts[index]`? 

What happens if we set `cur_inst` in the middle of a procedure? Should we instead check if `cur_inst == &otc_insts[index]` and in that case return `-EBUSY`?

Reply:
> Shouldn't this only be cleared if `cur_inst == &otc_insts[index]`?
> 
Now we only got "1" instance.

> What happens if we set `cur_inst` in the middle of a procedure? 
bt_ots_client_unregister should be only called when ACL disconnected.
Or I misunderstood the usage of this API. There is no documentation of this API.
> Should we instead check if `cur_inst == &otc_insts[index]` and in that case return `-EBUSY`?
My understand about cur_inst is coming from oacp_read(). I think the author using this pointer to prevent using OTS client API while previous operation is on-going.

Reply:
> Now we only got "1" instance.

Not sure what you mean by this, can you elaborate? 

> bt_ots_client_unregister should be only called when ACL disconnected.

But that is never checked or ensured, so if this is called while `cur_inst` is being used for a procedure, won't we just get NULL pointer access violation? Although it looks as though we have some (if not all) checks in places to verify that `cur_inst != NULL` before we attempt to use it.

Reply:
> > Now we only got "1" instance.
> 
> Not sure what you mean by this, can you elaborate?
> 
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/services/ots/ots_client.c#L35
OTS_CLIENT_INST_COUNT is hardcoded as 1.

> > bt_ots_client_unregister should be only called when ACL disconnected.
> 
> But that is never checked or ensured, so if this is called while `cur_inst` is being used for a procedure, won't we just get NULL pointer access violation? Although it looks as though we have some (if not all) checks in places to verify that `cur_inst != NULL` before we attempt to use it.

IMO, `memset(&otc_insts[index], 0, sizeof(otc_insts[index]))` will cause unwanted behavior while procedure is on-going or ACL disconnected. Is there anyway to checked if ots client is in middle of procedure?
This code snippet is not reliable if ACL is disconnected in middle of procedure.
`else if (inst->busy) {
		return -EBUSY;
	} else if (cur_inst) {
		return -EBUSY;
	}`

Reply:
> Is there anyway to checked if ots client is in middle of procedure? 

Yes, that's exactly what the `cur_inst` is for :) Is the instance currently in a procedure. So that's why we perhaps shouldn't set it to NULL here, because that would mean that we are interrupting an active procedure. 

I think it would be best to reject the call to `unregister` if `cur_inst != NULL`, if the instance being unregistered is the `cur_inst`.